### PR TITLE
In comm=ofi, use just one receive endpoint per locale instead of two.

### PR DESCRIPTION
**Pull request by gbtitus**
The ofi comm layer has long had two receive endpoints for each locale:
one for messaging (Active Messages, i.e. on-stmts) and one for RMA and
atomics.  We originally did this for load spreading.  We discovered
quite a while ago that such load spreading was unnecessary and in fact
that having two endpoints made things more complicated no matter whether
progress was automatic or manual, but the code was working and there
were more pressing things to deal with, so we left things as they were.
Here, drop back to just one receive endpoint per locale for all kinds of
traffic.

While here I also removed support for completion counters (vs.  queues)
for the receive endpoints.  For received AMs we need CQ events in order
to tell where the messages landed, and for received RMA and atomics we
don't make use of target-side events at all, so a receive CQ is the only
thing that makes sense.

This closes https://github.com/Cray/chapel-private/issues/2159.

### CODE REVIEW

**Comment by gbtitus** [comment]
@jhh67 I'm PR'ing this now mainly to clear the deck before my time off. I wouldn't want to merge it on a day when I wouldn't be working the next anyway. Let's plan on going through it together when I'm back on Monday the 28th, though of course feel free to look at it before then if you'd like.

**Comment by jhh67** [comment]
Will do.

John

From: Greg Titus ***@***.***>
Date: Wednesday, June 16, 2021 at 11:48 AM
To: chapel-lang/chapel ***@***.***>
Cc: Hartman, John ***@***.***>, Mention ***@***.***>
Subject: Re: [chapel-lang/chapel] In comm=ofi, use just one receive endpoint per locale instead of two. (#17924)

@jhh67<https://github.com/jhh67> I'm PR'ing this now mainly to clear the deck before my time off. I wouldn't want to merge it on a day when I wouldn't be working the next anyway. Let's plan on going through it together when I'm back on Monday the 28th, though of course feel free to look at it before then if you'd like.

File: runtime/src/comm/ofi/comm-ofi.c [hunk 1]
```diff
@@ -122,17 +122,11 @@ static struct fid_wait* ofi_amhWaitSet; // wait set for AM handler
 //
 static struct fid_ep* ofi_rxEp;         // AM req receive endpoint
```
Comment (gbtitus) [inline runtime/src/comm/ofi/comm-ofi.c:123]: fix comment

File: runtime/src/comm/ofi/comm-ofi.c [hunk 2]
```diff
@@ -6288,16 +6247,14 @@ void amEnsureProgress(struct perTxCtxInfo_t* tcip) {
     // Process the CQs/counters that had events.  We really only have
     // to take any explicit actions for our transmit endpoint.  If we
     // have inbound AM messages we want to handle those in the main
-    // poll loop.  And for the RMA endpoint we just need to ensure
-    // progress, which the poll call itself will have done.
+    // poll loop.  And for RMA we just need to ensure progress, which
```
Comment (gbtitus) [inline runtime/src/comm/ofi/comm-ofi.c:6250]: Look into doing this the same way as in the main loop.

File: runtime/src/comm/ofi/comm-ofi.c (hunk 1 above)
Comment (gbtitus) [reply runtime/src/comm/ofi/comm-ofi.c]: Done.

File: runtime/src/comm/ofi/comm-ofi.c (hunk 2 above)
Comment (gbtitus) [reply runtime/src/comm/ofi/comm-ofi.c]: Done.

**Comment by gbtitus** [comment]
Hi @jhh67 I think this is ready for you to take another look at. You could either look at the whole PR at once (again) using the `Files changed` button at the top, or look at the two commits I've added since our review this morning. I think the latter would be simpler, definitely. This is currently being full-tested on CS and EX systems.

**Review by jhh67 — APPROVED** [review]
Looks good to me.

**Pull request merged**